src/Compilers/CSharp/Portable/Symbols/Source/SourceNamespaceSymbol.cs

```diff
@@ -531,5 +542,707 @@ public void Add(NamespaceOrTypeSymbol symbol)
+                    if (!compilationUnit.Usings.Any())
+                    {
+                        return ImmutableDictionary<string, AliasAndUsingDirective>.Empty;
+                    }
+                    break;
+
+                case NamespaceDeclarationSyntax namespaceDecl:
+                    if (!namespaceDecl.Usings.Any())
+                    {
+                        return ImmutableDictionary<string, AliasAndUsingDirective>.Empty;
+                    }
+                    break;
+
+                default:
+                    throw ExceptionUtilities.UnexpectedValue(declarationSyntax);
+            }
+
+            foreach (var declaration in _mergedDeclaration.Declarations)
+            {
+                var declarationSyntaxRef = declaration.SyntaxReference;
+                if (declarationSyntaxRef.SyntaxTree != declarationSyntax.SyntaxTree)
+                {
+                    continue;
+                }
+
+                if (declarationSyntaxRef.GetSyntax() == declarationSyntax)
+                {
+                    return _aliasesAndUsings[declaration].GetUsingAliasesMap(this, declarationSyntax, basesBeingResolved);
+                }
+            }
+
+            Debug.Assert(false);
+            return ImmutableDictionary<string, AliasAndUsingDirective>.Empty;
+        }
+
+        public ImmutableArray<NamespaceOrTypeAndUsingDirective> GetUsingNamespacesOrTypes(CSharpSyntaxNode declarationSyntax, ConsList<TypeSymbol>? basesBeingResolved)
+        {
+            switch (declarationSyntax)
+            {
+                case CompilationUnitSyntax compilationUnit:
+                    if (!compilationUnit.Usings.Any())
+                    {
+                        return ImmutableArray<NamespaceOrTypeAndUsingDirective>.Empty;
+                    }
+                    break;
+
+                case NamespaceDeclarationSyntax namespaceDecl:
+                    if (!namespaceDecl.Usings.Any())
+                    {
+                        return ImmutableArray<NamespaceOrTypeAndUsingDirective>.Empty;
+                    }
+                    break;
+
+                default:
+                    throw ExceptionUtilities.UnexpectedValue(declarationSyntax);
+            }
+
+            foreach (var declaration in _mergedDeclaration.Declarations)
+            {
+                var declarationSyntaxRef = declaration.SyntaxReference;
+                if (declarationSyntaxRef.SyntaxTree != declarationSyntax.SyntaxTree)
+                {
+                    continue;
+                }
+
+                if (declarationSyntaxRef.GetSyntax() == declarationSyntax)
+                {
+                    return _aliasesAndUsings[declaration].GetUsingNamespacesOrTypes(this, declarationSyntax, basesBeingResolved);
+                }
+            }
+
+            Debug.Assert(false);
+            return ImmutableArray<NamespaceOrTypeAndUsingDirective>.Empty;
+        }
+
+        private class AliasesAndUsings
```

Comment:
Could we consider breaking this definition out into a separate partial class file? It feels a bit long and complex to have inline here.